### ISSUE TITLE
shutil.copytree is very slow on Windows

### ISSUE THREAD

**Issue by sztomi** [conan-io/conan#2170]
As evidenced by #2165 , there is plenty of room for improvement with shutil.copytree. With big packages, such as boost it takes a lot of time to copy sources to the build folder (even on very capable SSDs), and this adds up quickly with many packages and profiles to build. This is because the stdlib does not take advantage of async I/O and memory mapped files. This can be implemented in plain python in a portable way, but it is challenging. Failing that, a kind of extension point could offered to users to allow using e.g. `robocopy` to do the heavy lifting. This can lead to complex plugin architecturel discussion, which might have its place, but I think it's important to realize that users can already just monkeypatch the function, so a hypothetical plugin system should have a better value proposition than just that to be worth the effort.

**Comment by lasote** [comment]
As I commented in the #2165:

I would like to have this feature as:

A config entry in the conan.conf to opt-in the feature. system_copytree=False (opt-in to True).
Create an alternative copytree function using robocopy. It has to model the symlinks parameter too.
Wrap the shutil.copytree in a Conan function and, in windows use robocopy when it is activated (probably worth checking if present with tool.which).
Optionally, in Linux/Mac we could use the native cp that seems to already implement robocopy features. But in the initial version we could keep using the shutil.copytree and introduce the enhacement later.
I prefer to provide a solution for everyone than let the user declare the function in the conanfile.

**Comment by sztomi** [comment]
Modeling the symlinks parameter might be challenging (but possibly not necessary. I recall when I made my proof-of-concept, I found no need in the conan codebase to do it in the places my implementation replaced shutil.copytree).

On *nix platforms it might be worth adding [pyfastcopy](https://github.com/desbma/pyfastcopy) regardless of the setting as a fallback, although this only amends part of the reason why copying is slow.

As for future improvements, I think it's important to think ahead about how this could fit into a future extension system that was mentioned before. Even without defining such system in any way, it can be decided now whether an extension or this setting "wins" when they both want to change the same functionality (I think conan should bail with an error).
